sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java

```diff
@@ -0,0 +1,242 @@
+          + "called with null keyPattern");
+      return builder().setKeyPattern(keyPattern).build();
+    }
+
+    @Override
+    public void validate(PipelineOptions pipelineOptions) {
+      checkState(connectionConfiguration() != null,
+          "RedisIO.read() requires a connectionConfiguration to be set "
+              + "withConnection(connectionConfiguration)");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      if (connectionConfiguration() != null) {
+        connectionConfiguration().populateDisplayData(builder);
+      }
+    }
+
+    @Override
+    public PCollection<KV<String, String>> expand(PBegin input) {
+      return input
+          .apply(Create.of(keyPattern()))
+          .apply(RedisIO.readAll().withConnectionConfiguration(connectionConfiguration()));
+    }
+
+  }
+
+  /**
+   * Implementation of {@link #readAll()}.
+   */
+  @AutoValue
+  public abstract static class ReadAll
+      extends PTransform<PCollection<String>, PCollection<KV<String, String>>> {
+
+    @Nullable abstract RedisConnectionConfiguration connectionConfiguration();
+
+    abstract ReadAll.Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      @Nullable abstract ReadAll.Builder setConnectionConfiguration(
+          RedisConnectionConfiguration connection);
+      abstract ReadAll build();
+    }
+
+    public ReadAll withConnectionConfiguration(RedisConnectionConfiguration connection) {
+      checkArgument(connection != null, "RedisIO.read().withConnectionConfiguration"
+          + "(connectionConfiguration) called with null connectionConfiguration");
+      return builder().setConnectionConfiguration(connection).build();
+    }
+
+    @Override
+    public PCollection<KV<String, String>> expand(PCollection<String> input) {
+      return input
+          .apply(ParDo.of(
+              new ReadFn(connectionConfiguration())
+          ))
+          .setCoder(KvCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of()));
+      // reparallelized ?
+    }
+
+  }
+
+  /**
+   * A {@link DoFn} requesting Redis server to get key/value pairs.
+   */
+  private static class ReadFn extends DoFn<String, KV<String, String>> {
+
+    private final RedisConnectionConfiguration connectionConfiguration;
+
+    private transient Jedis jedis;
+
+    public ReadFn(RedisConnectionConfiguration connectionConfiguration) {
+      this.connectionConfiguration = connectionConfiguration;
+    }
+
+    @Setup
+    public void setup() {
+      jedis = connectionConfiguration.connect();
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext processContext) throws Exception {
+      ScanParams scanParams = new ScanParams();
+      scanParams.match(processContext.element());
+
+      String cursor = ScanParams.SCAN_POINTER_START;
+      boolean finished = false;
+      while (!finished) {
+        ScanResult<String> scanResult = jedis.scan(cursor, scanParams);
```

Comment:
I forget: does this scan need to be closed?

Reply:
No need, `ScanResult` doesn't have to be close (and it's an unsupported operation).